We will use a simple real-life example:

**👉 Marks → Result**



```
0 = Fail
1 = Pass
2 = Distinction

```



**🔹 STEP 1: Import the Machine Learning Tool**
```
from sklearn.linear_model import LogisticRegression
```

**Explanation:**

sklearn is a Python library

LogisticRegression is a ready-made machine learning model

We import it so we can use it

**👉 Just like importing math or random**

**🔹 STEP 2: Create a Small Dataset (Our Examples)**


```
X = [[40], [50], [60], [70], [80], [90]]

y = [0, 0, 1, 1, 2, 2]
```



**Explanation:**

X = input data (marks)

y = output labels (results)

Marks	Result

| Marks | Result          |
| ----- | --------------- |
| 40    | Fail (0)        |
| 50    | Fail (0)        |
| 60    | Pass (1)        |
| 70    | Pass (1)        |
| 80    | Distinction (2) |
| 90    | Distinction (2) |


👉 We are showing examples to the computer.

**🔹 STEP 3: Create the Model**
```
model = LogisticRegression(
    multi_class='multinomial',
    solver='lbfgs'
)
```

Explanation:

This creates a machine learning brain

multinomial → more than 2 classes

lbfgs → math method to learn patterns

👉 Right now the model knows nothing.

**🔹 STEP 4: Train (Teach) the Model**
```
model.fit(X, y)
```

Explanation:

The model looks at marks and results

Learns the pattern

This is the learning step

👉 This is where ML actually happens

**🔹 STEP 5: Ask the Model a Question**
```
prediction = model.predict([[65]])
print(prediction)
```
Explanation:

We ask: “What if marks = 65?”

The model gives an answer

Output is inside [ ] because it returns a list

**🔹 STEP 6: Ask for Confidence (Probability)**
```
probability = model.predict_proba([[65]])
print(probability)
```
Explanation:

This tells how confident the model is

It gives probability for each class

**🟢COMPLETE CODE🟢**

---



In [ ]:
from sklearn.linear_model import LogisticRegression

# Step 1: Dataset
X = [[40], [50], [60], [70], [80], [90]]
y = [0, 0, 1, 1, 2, 2]  # 0=Fail, 1=Pass, 2=Distinction

# Step 2: Create model
model = LogisticRegression(
    multi_class='multinomial',
    solver='lbfgs'
)

# Step 3: Train model
model.fit(X, y)

# Step 4: Predict result for 65 marks
print("Predicted Class:")
print(model.predict([[65]]))

# Step 5: Predict probabilities
print("\nPredicted Probabilities:")
print(model.predict_proba([[65]]))


Predicted Class:
[1]

Predicted Probabilities:
[[0.00348526 0.99302951 0.00348524]]


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(



```
Predicted Class:
[1]

Predicted Probabilities:
[[0.00348526 0.99302951 0.00348524]]

```

**🔹 [1] means:**
```
0 = Fail
1 = Pass
2 = Distinction
```

**👉 Prediction: PASS**


**🔹 Probabilities means:**
**[0.00348526 0.99302951 0.00348524]**

| Class | Meaning     | Probability |
| ----- | ----------- | ----------- |
| 0     | Fail        |      0.003    |
| 1     | Pass        |     0.99    |
| 2     | Distinction |       0.003   |

**👉 The model is most confident about PASS**

**🟢 WHY DO WE NEED POLYNOMIAL LOGISTIC REGRESSION?**

**First, recall NORMAL Logistic Regression:**

Draws a straight line to separate classes

Works only when data is linearly separable

❌ Problem:

Many real-world problems are curved, circular, or non-linear

Example:

Exam marks vs result

Age vs disease risk

Experience vs promotion

**👉 A straight line is not enough**

**🧠 IDEA BEHIND POLYNOMIAL LOGISTIC REGRESSION**

>“Don’t change the algorithm — change the FEATURES.”


We create new features like:

      3                2
    x                x

This allows logistic regression to:

**✔ Draw curved decision boundaries**

We will use a simple real-life example again:

🎯 Example:

Marks → Result
```
0 = Fail
1 = Pass
```


But now the relationship is non-linear

**🔹 STEP 1: Import Required Tools**
```
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
```
Explanation:

LogisticRegression → classifier

PolynomialFeatures → creates x², x³, etc.

👉 This is the key difference from normal logistic regression

**🔹 STEP 2: Create Dataset**
```
X = [[20], [30], [40], [50], [60], [70], [80]]
y = [0, 0, 0, 1, 1, 1, 0]
```
Explanation:

Marks	Result

| Marks | Result |
| ----- | ------ |
| 20    | Fail   |
| 30    | Fail   |
| 40    | Fail   |
| 50    | Pass   |
| 60    | Pass   |
| 70    | Pass   |
| 80    | Fail   |


⚠️ Notice:

High marks (80) → Fail

This cannot be solved by a straight line

🔹 STEP 3: Convert Marks into Polynomial Features
```
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
```
What happens here?

If input is:
```
[50]
```

After polynomial transform:
```
[1, 50, 2500]
```

Meaning:

- 1 (bias)

- x

- x²

👉 This allows the model to bend the line

**🔹 STEP 4: Create Logistic Regression Model**
```
model = LogisticRegression()
```
Explanation:

Same logistic regression

Nothing new here

Magic is in features, not model

**🔹 STEP 5: Train the Model**
```
model.fit(X_poly, y)
```
Explanation:

Model learns pattern using x and x²

This creates a curved decision boundary

**🔹 STEP 6: Make Prediction**
```
test_mark = [[55]]
test_poly = poly.transform(test_mark)

print(model.predict(test_poly))
print(model.predict_proba(test_poly))
```
Explanation:

Convert test data same way

Predict class and confidence

**🟢COMPLETE CODE🟢**

---



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures

# Step 1: Dataset
X = [[20], [30], [40], [50], [60], [70], [80]]
y = [0, 0, 0, 1, 1, 1, 0]  # 0=Fail, 1=Pass

# Step 2: Polynomial feature generator
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Step 3: Create model
model = LogisticRegression()

# Step 4: Train model
model.fit(X_poly, y)

# Step 5: Test with new value
test_mark = [[55]]
test_poly = poly.transform(test_mark)

print("Predicted Class:")
print(model.predict(test_poly))

print("\nPredicted Probabilities:")
print(model.predict_proba(test_poly))


Predicted Class:
[1]

Predicted Probabilities:
[[0.09126635 0.90873365]]


**[1] means:**
```
0 = Fail
1 = Pass
```

👉 Prediction: PASS



**Predicted Probabilities:**
**[[0.09126635 0.90873365]]**

| Class | Meaning | Probability |
| ----- | ------- | ----------- |
| 0     | Fail    |         |
| 1     | Pass    |         |

👉 The model is confident that the student passes